ctk/common/src/main/kotlin/org/codice/compliance/verification/core/SamlDefinedIdentifiersVerifier.kt

```diff
@@ -118,4 +130,48 @@ internal class SamlDefinedIdentifiersVerifier(val node: Node) {
             }
         }
     }
+
+    /** 8.3.7 Persistent Identifier */
+    private fun verifyPersistentIdentifiers() {
+        node.recursiveChildren()
+            .filter { it.namespaceURI != null }
+            .filter { it.namespaceURI == PERSISTENT_ID }
+            .forEach {
+                if (it.textContent != null && it.textContent.length > ID_VALUE_LENGTH_LIMIT)
+                    throw SAMLComplianceException.create(SAMLCore_8_3_7_a,
+                        message = "The length of the Persistent ID's value was greater than " +
+                            "$ID_VALUE_LENGTH_LIMIT characters.",
+                        node = it)
+
+                if (it.children("NameQualifier")
+                        .any { it.textContent != idpMetadata.entityId })
+                    throw SAMLComplianceException.create(SAMLCore_8_3_7_b,
+                        SAMLCore_8_3_7_c,
+                        message = "The Persistent ID's NameQualifier isn't ${idpMetadata.entityId}",
+                        node = it)
+
+                if (it.children("SPNameQualifier")
+                        .any { it.textContent != SP_ISSUER })
+                    throw SAMLComplianceException.create(SAMLCore_8_3_7_d,
+                        message = "The Persistent ID's SPNameQualifier isn't $SP_ISSUER",
+                        node = it)
+            }
+    }
+
+    /** 8.3.8 Transient Identifier */
+    private fun verifyTransientIdentifiers() {
+        node.recursiveChildren()
+            .filter { it.namespaceURI != null }
+            .filter { it.namespaceURI == TRANSIENT_ID }
+            .filter { it.textContent != null }
+            .forEach {
+                if (it.textContent.length > ID_VALUE_LENGTH_LIMIT)
+                    throw SAMLComplianceException.create(SAMLCore_8_3_8_a,
+                        message = "The length of the Transient ID's value was greater than " +
+                            "$ID_VALUE_LENGTH_LIMIT characters.",
+                        node = it)
+
+                CommonDataTypeVerifier.verifyIdValues(it, SAMLCore_8_3_8_a)
```

Comment:
❓ Is there a reason we're not making the `CommonDataType` call for persistent identifiers?

Reply:
That's because the Transient explicitly says: `Transient identifier values MUST be generated in accordance with the rules for SAML identifiers (see Section 1.3.4)`